client/opentracing/src/main/java/org/hawkular/apm/client/opentracing/TraceContext.java

```diff
@@ -61,25 +61,31 @@
 
     private TraceRecorder recorder;
     private ContextSampler sampler;
+    private Property serviceName;
 
     private static List<NodeProcessor> nodeProcessors = new ArrayList<>();
 
     static {
         nodeProcessors.add(new DefaultNodeProcessor());
     }
 
+    private static Property ENV_SERVICE_NAME = new Property(Constants.PROP_SERVICE_NAME, getServiceName());
+    private static Property ENV_BUILD_STAMP = new Property(Constants.PROP_BUILD_STAMP, getBuildStamp());
+
     /**
      * This constructor initialises the trace context.
      *
      * @param topSpan The top level span in the trace
      * @param rootNode The builder for the root node of the trace fragment
      * @param recorder The trace recorder
+     * @param serviceName The service name to be assigned to the root node, or null if a service name should be guessed
      */
-    public TraceContext(APMSpan topSpan, NodeBuilder rootNode, TraceRecorder recorder, ContextSampler sampler) {
+    public TraceContext(APMSpan topSpan, NodeBuilder rootNode, TraceRecorder recorder, ContextSampler sampler, Property serviceName) {
```

Comment:
Sorry slightly misunderstood the question - yes it would be good to also pass in the constructor, so the application as the ability to directly provide the service name and build details rather than have it derived from the environment, or explicitly provided as a tag on each top level span.